systemdmanager.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2020 Open Mobile Platform LLC.
+ *
+ * All rights reserved.
+ *
+ * BSD 3-Clause License, see LICENSE.
+ */
+
+#include "systemdmanager.h"
+#include "logging.h"
+#include <QDBusInterface>
+#include <QDBusObjectPath>
+#include <QDBusPendingCall>
+#include <QDBusPendingReply>
+
+namespace Systemd {
+const auto Service = QStringLiteral("org.freedesktop.systemd1");
+const auto ManagerPath = QStringLiteral("/org/freedesktop/systemd1");
+const auto ManagerInterface = QStringLiteral("org.freedesktop.systemd1.Manager");
+const auto Replace = QStringLiteral("replace");
+const auto StartUnit = QStringLiteral("StartUnit");
+const auto StopUnit = QStringLiteral("StopUnit");
+const auto ResultDone = QStringLiteral("done");
+const auto ResultSkipped = QStringLiteral("skipped");
+}
+
+// This is currently implemented so that it can do one thing at a time
+// as there is no need for anything more complicated.
+// It would be quite easily possible to add individual queues with
+// queue numbers if needed.
+
+SystemdManager::SystemdManager(QObject *parent) :
+    QObject(parent),
+    m_systemd(new QDBusInterface(Systemd::Service, Systemd::ManagerPath,
+                                 Systemd::ManagerInterface,
+                                 QDBusConnection::systemBus(), this))
+{
+    if (!m_systemd->isValid())
+        qCCritical(lcSUM) << "Could not create interface to systemd, can not function!";
+    if (!connect(m_systemd, SIGNAL(JobRemoved(uint, QDBusObjectPath, QString, QString)),
+                 this, SLOT(onJobRemoved(uint, QDBusObjectPath, QString, QString))))
+        qCCritical(lcSUM) << "Could not connect to JobRemoved signal, can not function!";
+}
+
+SystemdManager::~SystemdManager()
+{
+    delete m_systemd;
+    m_systemd = nullptr;
+}
+
+bool SystemdManager::busy()
+{
+    // Busy if there is something on queue or a pending call or job removal is waited for
+    return !m_jobs.isEmpty() || m_pendingCall || !m_currentJob.isEmpty();
+}
+
+void SystemdManager::addUnitJob(Job job)
+{
+    addUnitJobs(JobList() << job);
+}
+
+void SystemdManager::addUnitJobs(JobList &jobs)
+{
+    Q_ASSERT_X("addUnitJobs", !jobs.isEmpty(), "jobs must never be empty");
+    bool wasEmpty = m_jobs.isEmpty();
+    m_jobs.append(jobs);
+    processNextJob();
+    if (wasEmpty)
+        emit busyChanged();
+}
+
+void SystemdManager::processNextJob()
+{
+    if (m_pendingCall || !m_currentJob.isEmpty())
+        return;
+
+    qCDebug(lcSUM) << "Process next systemd job";
+
+    QDBusPendingCall call = m_systemd->asyncCall(
+            (m_jobs.first().type == StopJob) ? Systemd::StopUnit : Systemd::StartUnit,
+            m_jobs.first().unit, Systemd::Replace);
+    QDBusPendingCallWatcher *m_pendingCall = new QDBusPendingCallWatcher(call, this);
+    connect(m_pendingCall, &QDBusPendingCallWatcher::finished, this, &SystemdManager::onJobStarted);
```

Comment:
`SystemdManager::onJobStarted` is invoked even if job type was `StopJob`, right?

Reply:
Job != Unit. So in systemd speak job is a task of systemd and thus starting or stopping a unit is a job. So yes, it's correct to invoke that on both types of job.

Reply:
Yeah, I solved that riddle. Nice way to confuse the reader!

Reply:
But it's basically a queue of D-Bus calls, and a "job" is a D-Bus call, right? It's just being referred to as a "job".

Reply:
I'm not sure if I follow. So for each (systemd) job we have a D-Bus call to start it and then a signal that tell that the job has ended (to a failure or success). We need to handle both of these.

Reply:
If onJobStarted was onCallFinished , and m_jobs was m_callQueue or something like this, the code would be significantly easier to understand. At least to me.

When I saw "job" I was thinking that it was another name for a system unit. But in the end it turned out to be another name for a D-Bus call. That was confusing.

Basically, I suggest to refer to D-Bus calls as "D-Bus calls", that's all. Unless the goal is to confuse the reader. In that case - good job, you have confused me!

Reply:
Well of course my goal is to confuse you. :roll_eyes: No need to be so hostile about this. I think you are right that I could probably name onJobStarted better. Still I think these are not D-Bus calls as such but instead they are systemd jobs and involve more than just a D-Bus call. Note that I'm only removing them from queue once the job is done. But I can still reconsider the naming and add some explanatory comments if they seem useful.

Reply:
oh come on, "good job" was a calambur, no hostility intended or anything. There were systemd-style job control systems on other OSes where units were called jobs, that experience of mine contributed to confusion.

Reply:
In other words, feel free to ignore.